File: server/routes/ai.py
import sqlite3

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import JSONResponse

from auth import get_current_user
from database import get_db
from limiter import limiter
from models import CategorizeRequest
from services.ai_service import generate_ai_insights, categorize_item_with_ai, generate_portfolio_summary
from services.fallback_service import generate_fallback_insights

router = APIRouter()


def row_to_dict(row: sqlite3.Row) -> dict:
    return dict(row)


@router.post("/insights/{item_id}")
@limiter.limit("30/15minutes")
def get_insights(
    request: Request,
    item_id: int,
    mode: str = Query("ai"),
    db: sqlite3.Connection = Depends(get_db),
    _: str = Depends(get_current_user),
):
    row = db.execute("SELECT * FROM inventory WHERE id = ?", [item_id]).fetchone()
    if not row:
        return JSONResponse(status_code=404, content={"error": f"Item {item_id} not found"})
    item = row_to_dict(row)
    if mode == "rule":
        return generate_fallback_insights(item)
    return generate_ai_insights(item)


@router.post("/bulk-insights")
@limiter.limit("30/15minutes")
def bulk_insights(request: Request, db: sqlite3.Connection = Depends(get_db), _: str = Depends(get_current_user)):
    rows = db.execute("SELECT * FROM inventory").fetchall()
    return [
        {"item": row_to_dict(row), "insight": generate_fallback_insights(row_to_dict(row))}
        for row in rows
    ]


@router.post("/portfolio-summary")
@limiter.limit("10/15minutes")
def portfolio_summary(request: Request, db: sqlite3.Connection = Depends(get_db), _: str = Depends(get_current_user)):
    rows = db.execute("SELECT * FROM inventory").fetchall()
    items = [row_to_dict(r) for r in rows]
    return generate_portfolio_summary(items)


@router.post("/categorize")
@limiter.limit("30/15minutes")
def categorize(
    request: Request,
    data: CategorizeRequest,
    db: sqlite3.Connection = Depends(get_db),
    _: str = Depends(get_current_user),
):
    if not data.name or not data.name.strip():
        return JSONResponse(status_code=400, content={"error": "Name is required for categorization"})
    return {"category": categorize_item_with_ai(data.name, data.notes or "")}

File: server/src/routes/ai.ts
import { Router, Request, Response } from 'express';
import { Database } from 'better-sqlite3';
import { InventoryItem } from '../db';
import { generateAIInsights, categorizeItemWithAI } from '../services/aiService';
import { generateFallbackInsights } from '../services/fallbackService';

export function createAIRouter(db: Database): Router {
  const router = Router();

  // POST /insights/:id — get AI insights for a specific item
  router.post('/insights/:id', async (req: Request, res: Response) => {
    try {
      const id = parseInt(req.params.id, 10);
      if (isNaN(id)) {
        return res.status(400).json({ error: 'Invalid item ID' });
      }

      const item = db.prepare('SELECT * FROM inventory WHERE id = ?').get(id) as InventoryItem | undefined;
      if (!item) {
        return res.status(404).json({ error: `Item with ID ${id} not found` });
      }

      const allItems = db.prepare('SELECT * FROM inventory').all() as InventoryItem[];
      const insights = await generateAIInsights(item, allItems);
      return res.json(insights);
    } catch (error) {
      console.error('POST /ai/insights/:id error:', error);
      return res.status(500).json({ error: 'Failed to generate insights' });
    }
  });

  // POST /bulk-insights — get fallback insights for all items
  router.post('/bulk-insights', (_req: Request, res: Response) => {
    try {
      const allItems = db.prepare('SELECT * FROM inventory').all() as InventoryItem[];
      const results = allItems.map(item => {
        const insight = generateFallbackInsights(item);
        return {
          item,
          insight,
        };
      });
      return res.json(results);
    } catch (error) {
      console.error('POST /ai/bulk-insights error:', error);
      return res.status(500).json({ error: 'Failed to generate bulk insights' });
    }
  });

  // POST /categorize — suggest a category for an item using AI
  router.post('/categorize', async (req: Request, res: Response) => {
    try {
      const { name, notes } = req.body;

      if (!name || typeof name !== 'string' || name.trim() === '') {
        return res.status(400).json({ error: 'Item name is required for categorization' });
      }

      const category = await categorizeItemWithAI(
        name.trim(),
        typeof notes === 'string' ? notes.trim() : ''
      );

      return res.json({ category });
    } catch (error) {
      console.error('POST /ai/categorize error:', error);
      return res.status(500).json({ error: 'Failed to categorize item', category: 'Office Supplies' });
    }
  });

  return router;
}
